src/main/java/org/apache/ibatis/builder/xml/XMLIncludeTransformer.java

```diff
@@ -37,9 +41,42 @@ public XMLIncludeTransformer(Configuration configuration, MapperBuilderAssistant
   }
 
   public void applyIncludes(Node source) {
+    Properties variablesContext = new Properties();
+    Properties configurationVariables = configuration.getVariables();
+    if (configurationVariables != null) {
+      variablesContext.putAll(configurationVariables);
+    }
+    applyIncludes(source, variablesContext);
+  }
+
+  /**
+   * Recursively apply includes through all SQL fragments.
+   * @param source Include node in DOM tree
+   * @param variablesContext Current context for static variables with values
+   */
+  private void applyIncludes(Node source, final Properties variablesContext) {
     if (source.getNodeName().equals("include")) {
-      Node toInclude = findSqlFragment(getStringAttribute(source, "refid"));
-      applyIncludes(toInclude);
+      // new full context for included SQL - contains inherited context and new variables from current include node
+      Properties fullContext;
+
+      String refid = getStringAttribute(source, "refid");
+      // replace variables in include refid value
+      refid = PropertyParser.parse(refid, variablesContext);
+      Node toInclude = findSqlFragment(refid);
+      Properties newVariablesContext = getVariablesContext(source);
+      if (!newVariablesContext.isEmpty()) {
+        // replace variables in variable values too
+        for (Object name : newVariablesContext.keySet()) {
+          newVariablesContext.put(name, PropertyParser.parse(newVariablesContext.get(name).toString(), variablesContext));
+        }
+        // merge new and inherited into new full one
+        applyInheritedContext(newVariablesContext, variablesContext);
+        fullContext = newVariablesContext;
+      } else {
+        // no new context - use inherited fully
+        fullContext = variablesContext;
+      }
+      applyIncludes(toInclude, fullContext);
```

Comment:
This part could be simpler if replacing variables in the subroutine.
What do you think?

```java
Properties fullContext;
Properties childContext = getVariablesContext(source, variablesContext);
if (childContext.isEmpty()) {
  fullContext = variablesContext;
} else {
  fullContext = new Properties();
  fullContext.putAll(variablesContext);
  fullContext.putAll(childContext);
}
applyIncludes(toInclude, fullContext);
```

I could be missing something, though.